src/EditorFeatures/Core/ValueTracking/ValueTrackingService.Visitor.cs

```diff
@@ -0,0 +1,266 @@
+                    return AddReference(operation, cancellationToken);
+                }
+
+                return Task.CompletedTask;
+
+                Task AddReference(IOperation operation, CancellationToken cancellationToken)
+                    => operation switch
+                    {
+                        IParameterReferenceOperation parameterReference => AddOperationAsync(operation, parameterReference.Parameter, cancellationToken),
+                        IFieldReferenceOperation fieldReferenceOperation => AddOperationAsync(operation, fieldReferenceOperation.Member, cancellationToken),
+                        IPropertyReferenceOperation propertyReferenceOperation => AddOperationAsync(operation, propertyReferenceOperation.Member, cancellationToken),
+                        ILocalReferenceOperation localReferenceOperation => AddOperationAsync(operation, localReferenceOperation.Local, cancellationToken),
+                        _ => Task.CompletedTask
+                    };
+            }
+
+            private Task VisitLiteralAsync(ILiteralOperation literalOperation, CancellationToken cancellationToken)
+            {
+                if (literalOperation.Type is null)
+                {
+                    return Task.CompletedTask;
+                }
+
+                return AddOperationAsync(literalOperation, literalOperation.Type, cancellationToken);
+            }
+
+            private Task VisitReturnAsync(IReturnOperation returnOperation, CancellationToken cancellationToken)
+            {
+                if (returnOperation.ReturnedValue is null)
+                {
+                    return Task.CompletedTask;
+                }
+
+                return VisitAsync(returnOperation.ReturnedValue, cancellationToken);
+            }
+
+            private async Task AddOperationAsync(IOperation operation, ISymbol symbol, CancellationToken cancellationToken)
+            {
+                _ = await ProgressCollector.TryReportAsync(
+                        Solution,
+                        operation.Syntax.GetLocation(),
+                        symbol,
+                        cancellationToken: cancellationToken).ConfigureAwait(false);
+            }
+
+            private async Task TrackArgumentsAsync(ImmutableArray<IArgumentOperation> argumentOperations, CancellationToken cancellationToken)
+            {
+                var collectorsAndArgumentMap = argumentOperations
+                    .Where(ShouldTrackArgument)
+                    // Clone the collector here to allow each argument to report multiple items.
+                    // See Clone() docs for more details
+                    .Select(argument => (collector: Clone(), argument))
+                    .ToImmutableArray();
+
+                var tasks = collectorsAndArgumentMap
+                    .Select(pair => Task.Run(() => pair.collector.VisitAsync(pair.argument, cancellationToken)));
```

Comment:
we should not do parallel work in proc.  this can totally saturate the GC/CPU and really affect latency.  If you're in OOP, this is fine :)

Reply:
https://github.com/dotnet/roslyn/issues/52435 as follow up to making this service OOP